### PR TITLE
semi-colon, not colon

### PR DESCRIPTION
**Pull request by sidewayss**
Regarding this section of the page:

> The name must not start with xml (case-insensitive).
> The name must not contain any colon characters (:).
> The name must not contain any capital letters.

In fact, colons are allowed, semi-colons are not.  This matches the [SVG docs page](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/data-*) too.

Here are some running examples of how this behaves, choose your preferred site and read the console messages for the results.  These examples illustrate the behavior behind this PR's fix and some additional behaviors summarized below:
[mdn](https://developer.mozilla.org/en-US/play?id=BRuAYNK%2BsT3BYBardgrlWcDvmCbWZEE%2FW0C0MXjHQv6%2F%2BI6gWcaHth%2BvIE1GPAksPkqD6ethDg%2F0hSQI)
https://codepen.io/sidewayss/pen/dywPGZa
https://jsfiddle.net/sidewayss/oz2vx6rc/15/

On Chrome and Firefox I get the same results.  Both "xml" and "data-xml" attributes can be set on an `HTMLElement`.   Upper-case alpha characters are not prohibited, they are forced to lower-case, and `HTMLElement.dataset` enforces the use of lower-case when getting a value.  `get|setAttribute()` are truly case-insensitive, they accept any case for any letter. The `HTMLElement.dataset` snake-case to/from camelCase conversion has some issues with '--' and '-' followed by a number.

It's not as simple as 3 bullet points, unfortunately, and the "xml" exclusion might be something in the spec, but you can see for yourself that it's not enforced.  The SVG version of this page is also incorrect on these two points.  And, of course, if you want your attribute name to work with `HTMLElement.dataset` and `.property` vs `["property"]` it must also conform to JavaScript limitations: alpha-numeric and underscore only, cannot start with a number. 

So here is this one fix as a PR, but please let me know if/how you'd like to pursue correcting the other two bullet points.

### CODE REVIEW

**Comment by estelle** [comment]
My notes while looking at this:

The spec at https://html.spec.whatwg.org/multipage/dom.html#attr-data-* reads:

> A custom data attribute is an attribute in no namespace whose name starts with the string "data-", has at least one character after the hyphen, is [XML-compatible](https://html.spec.whatwg.org/multipage/infrastructure.html#xml-compatible), and contains no [ASCII upper alphas](https://infra.spec.whatwg.org/#ascii-upper-alpha).

(The attribute names on [HTML elements](https://html.spec.whatwg.org/multipage/infrastructure.html#html-elements) in [HTML documents](https://dom.spec.whatwg.org/#html-document) get ASCII-lowercased automatically, so the restriction on ASCII uppercase letters doesn't affect such documents.)

The names need to be XML compatible. https://html.spec.whatwg.org/multipage/infrastructure.html#xml-compatible reads:

> Attribute names are said to be XML-compatible if they match the [Name](https://www.w3.org/TR/xml/#NT-Name) production defined in XML and they contain no U+003A COLON characters (:). [[XML]](https://html.spec.whatwg.org/multipage/references.html#refsXML)

The unicode characters are:

unicode code point | character | UTF-8 (hex.) | name |
-- | -- | -- | --
U+003A | : | 3a | COLON
U+003B | ; | 3b | SEMICOLON

MDN documentation contradicts itself.  https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset reads:

> The attribute name begins with data-. It can contain only letters, numbers, dashes (-), periods (.), colons (:), and underscores (_). Any ASCII capital letters (A to Z) are converted to lowercase.

In the codepen, there are some things that i wondered about; not sure if those were proving an issue or coding issues. Here's my version: https://codepen.io/estelle/pen/PoXweMZ

**Comment by estelle** [comment]
Deeper diving:

The text reads:

> The * may be replaced by any name following [the production rule of XML names](https://www.w3.org/TR/REC-xml/#NT-Name) with the following restrictions:

it likely should read something like this:

> The * may be replaced by any name following [the production rule of XML names](https://www.w3.org/TR/REC-xml/#NT-Name) which includes the following restrictions:

and then change all the "must not"s to "should not"s. Because they shouldn't. That doesn't mean they can't.

We should then add a paragraph about how, 

The XML spec contradicts the list, kind of....

In the XML spec, it basically says "colons should not be used but need to be parsed anyhow"

The Namespaces in XML Recommendation [[XML Names]](https://www.w3.org/TR/xml/#xml-names) assigns a meaning to names containing colon characters. Therefore, authors should not use the colon in XML names except for namespace purposes, but XML processors must accept the colon as a name character.

See the two paragraphs above https://www.w3.org/TR/xml/#d0e804 and the tokens below:

> Document authors are encouraged to use names which are meaningful words or combinations of words in natural languages, and to avoid symbolic or white space characters in names. Note that COLON, HYPHEN-MINUS, FULL STOP (period), LOW LINE (underscore), and MIDDLE DOT are explicitly permitted.

> The ASCII symbols and punctuation marks, along with a fairly large group of Unicode symbol characters, are excluded from names because they are more useful as delimiters in contexts where XML names are used outside XML documents; providing this group gives those contexts hard guarantees about what cannot be part of an XML name. The character #x037E, GREEK QUESTION MARK, is excluded because when normalized it becomes a semicolon, which could change the meaning of entity references.

**Comment by sidewayss** [comment]
I like your version of the codepen.  It doesn't invalidate anything in the original.  For some of it I was simply testing different possibilities that I thought might be problematic, given my experience switching between snake and camel case and dealing with JavaScript naming conventions.  Numbers and extra hyphens are common gotchas.  I didn't even try `.` because that's definitely not going to work with `dataset.property` syntax.

Regarding your new comment:
- My only experience with colons is the now deprecated `xlink:href` version of the SVG `href` attribute.  There it indicates the namespace `xlink`.
- If you want your attribute name to work with `dataset.property` syntax, JavaScript defines additional restrictions, which should certainly be noted in the `HTMLElement.dataset` docs.
- I agree about the "should" vs "must" rephrasing, see below.

So how do you wish to proceed?  I am proposing to approve this minimal change and proceed to fix the rest in a separate PR, hopefully authored by you :-).  I can take a crack at it if you prefer.

In practice, though they may be forbidden in the spec, colons are completely usable, as is the string "xml" anywhere in the name.  You can also name any attribute almost anything, so the "data-" prefix is a best-practice that avoids name collision with existing attributes and collects the custom attribute in `dataset` along with some convenient, though flawed, camelCase conversion, which is useful, but not required.  How would you like to approach those issues?  Or should I just hand it all over to you now?

**Comment by estelle** [comment]
I also suggest editing the paragraph after the list.

> Note that the [HTMLElement.dataset](/en-US/docs/Web/API/HTMLElement/dataset) property is a [DOMStringMap](/en-US/docs/Web/API/DOMStringMap), and the name of the custom data attribute data-test-value will be accessible via HTMLElement.dataset.testValue (or by HTMLElement.dataset["testValue"]) as any dash (U+002D) is replaced by the capitalization of the next letter, converting the name to camelcase.

To something similar to (please rewrite this as appropriate):

> The [HTMLElement.dataset](/en-US/docs/Web/API/HTMLElement/dataset) property is a [DOMStringMap](/en-US/docs/Web/API/DOMStringMap). When accessing `data-*` attributes using the `dataset` property using JavaScript, the name of the custom data attribute is converted to a camelCased keyname. The `data-` prefix is removed, the letter following each hyphen is converted to uppercase, and other characters stay unchanged. For example, `data-test-value` will be accessible via `HTMLElement.dataset.testValue` (or by HTMLElement.dataset["testValue"]) as any dash (U+002D) is replaced by the capitalization of the next letter, converting the name to camelcase. Therefore, any character coming after a hyphen that doesn't have an uppercase equivalent is not valid.  And any character that is not valid within a variable name in JavaScript will cause errors when used with `dataset.property` syntax.

**Comment by estelle** [comment]
we were both commenting at the same time.

My suggestion:

1) 
replace:

> The * may be replaced by any name following [the production rule of XML names](https://www.w3.org/TR/REC-xml/#NT-Name) with the following restrictions:

with:

> The * may be replaced by any name following [the production rule of XML names](https://www.w3.org/TR/REC-xml/#NT-Name) which includes the following restrictions:

2)

 change the three "must not"s to "should not"s. Because they shouldn't, and that is what the spec kind of says, 

and 

3) change the paragraph following the list from:

> Note that the [HTMLElement.dataset](/en-US/docs/Web/API/HTMLElement/dataset) property is a [DOMStringMap](/en-US/docs/Web/API/DOMStringMap), and the name of the custom data attribute data-test-value will be accessible via HTMLElement.dataset.testValue (or by HTMLElement.dataset["testValue"]) as any dash (U+002D) is replaced by the capitalization of the next letter, converting the name to camelcase.

To a version of the following, but explained more pedantically:

The [HTMLElement.dataset](/en-US/docs/Web/API/HTMLElement/dataset) property is a [DOMStringMap](/en-US/docs/Web/API/DOMStringMap). When accessing `data-* ` using the dataset property using JavaScript, the name of the custom data attribute is converted to a camelCased keyname. The data- prefix is removed, the letter following each hyphen is converted to uppercase, and other characters stay unchanged. For example, `data-test-value` will be accessible via `HTMLElement.dataset.testValue` (or by `HTMLElement.dataset["testValue"]) `as any dash (`U+002D`) is replaced by the capitalization of the next letter, converting the name to camelcase. Therefore, any character coming after a hyphen that doesn't have an uppercase equivalent is not valid. And any character that is not valid within a variable name in JavaScript will cause errors when used with `dataset.property` syntax.

i am heading out for an hour, so promise not to overlap any comments during that time ;)

**Comment by sidewayss** [comment]
I think I should also double up the colon bullet point with semi-colon.  Neither should be used and one will throw an error when you try to use it.  They are already grouped together in Unicode, as you pointed out.

**Comment by sidewayss** [comment]
I forgot `$` as a valid JavaScript variable/property character, probably because no one uses it except for JQuery and it's only allowed as the first character.  It is not allowed at all in an attribute name, so the two sets of limitations don't totally overlap.

**Comment by estelle** [comment]
>. I haven't tried using CSS, though it uses semi-colons in its syntax too, so probably not ok there.
> 
> You can create the attribute but you can't access its value.

### HTML/CSS and allowed characters

I assumed HTML only balked at a few characters, like space, equal, quotes, `/`,`&`, `>`, and `<`; basically characters that have meaning within HTML elements. CSS has supported all of these as long as they are escaped. I can't find the post i wrote, but it's likely outdated as there was a specific resolution https://github.com/w3c/csswg-drafts/issues/7129#issuecomment-1069331981. Conveniently, the thread lists the HTML chars that are allowed https://github.com/w3c/csswg-drafts/issues/7129#issuecomment-1069337262

### Colon v. semi-colon

The specification specifically reads "and they contain no U+003A COLON characters (:)." so we can't go with the edit you suggested as is, because even though it is what is supported, the sentence before it reads "[the production rule of XML names]" which lists colon, not semi-colon. so we can't change one without changing the other.

**Comment by sidewayss** [comment]
> The specification specifically reads "and they contain no U+003A COLON characters (:)." so we can't go with the edit you suggested as is, because even though it is what is supported, the sentence before it reads "[the production rule of XML names]" which lists colon, not semi-colon. so we can't change one without changing the other.

Why not eliminate "[the production rule of XML names]" from that sentence and list the actual rules?

**Comment by sidewayss** [comment]
> CSS has supported all of these as long as they are escaped.

Sorry, I'm much more focused on the javascript side of this and tend to forget the CSS part.  So every and any attribute you can create in HTML can be accessed by CSS? It doesn't need to start with "data-" or any other restrictions?  Sounds like javascript is the difficult one here, but `HTMLElement.dataset` puts it in the center of the frame.  Do you have any idea how these custom attributes are used globally, more via CSS or JavaScript?

As for the resolution to this, I'm fine with your suggestions as-is.  I'm in the middle of something that touched on this and I took a detour here. I will try to make the edits to this page tomorrow or Wed at the latest.  I want to clear my head of my current problem first before I get "more pedantic".  Thanks.

**Comment by estelle** [comment]
Basically, this is an HTML page. So we have to focus on the "HTML" and other XML base markup language rules on this page. That is why the XML sentence is there. The CSS and JS bits are "asides" on this page. In other words, the main content needs to be what the WHATWG (and to a lesser extent XML and SVG) docs say, with "caveats" for CSS and JS. 

The WHATWG discusses the XML syntax.  That is why it is there. I think the [WHATWG definition]( https://html.spec.whatwg.org/multipage/dom.html#attr-data-*) and the [comment about characters allowed in HTML attributes](https://github.com/w3c/csswg-drafts/issues/7129#issuecomment-1069337262) is what we should be citing, then adding a note or paragraph explaining the issues with JS.

**Comment by sidewayss** [comment]
> The WHATWG discusses the XML syntax. That is why it is there. I think the [WHATWG definition](https://html.spec.whatwg.org/multipage/dom.html#attr-data-*) and the [comment about characters allowed in HTML attributes](https://github.com/w3c/csswg-drafts/issues/7129#issuecomment-1069337262) is what we should be citing, then adding a note or paragraph explaining the issues with JS.

As in replacing the [the production rule of XML names](https://www.w3.org/TR/REC-xml/#NT-Name) link right before the 3 bullets with these two links and a bit of rewording?

I have implemented your other suggestions and pushed the changes here.

File: files/en-us/web/html/global_attributes/data-_star_/index.md [hunk 1]
```diff
@@ -12,13 +12,13 @@ The **`data-*`** [global attributes](/en-US/docs/Web/HTML/Global_attributes) for
...
 All such custom data are available via the {{domxref("HTMLElement")}} interface of the element the attribute is set on. The {{domxref("HTMLElement.dataset")}} property gives access to them.
-The `*` may be replaced by any name following [the production rule of XML names](https://www.w3.org/TR/REC-xml/#NT-Name) with the following restrictions:
+The `*` may be replaced by any name following [the production rule of XML names](https://www.w3.org/TR/REC-xml/#NT-Name) which includes the following restrictions:
 
-- The name must not start with `xml` (case-insensitive).
-- The name must not contain any colon characters (`:`).
-- The name must not contain any capital letters.
+- The name should not start with `xml` (case-insensitive).
+- The name should not contain any semi-colon characters (`;`).
+- The name should not contain any capital letters.
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/data-_star_/index.md:19]:
```suggestion
- The name should not start with `xml` (case-insensitive), as it's reserved for future XML specifications.
- The name should not contain any colon characters (`:`), as XML assigns meaning to such names.
- The name should not contain any capital letters, as XML is all lowercase.
```

File: files/en-us/web/html/global_attributes/data-_star_/index.md [hunk 2]
```diff
@@ -12,13 +12,13 @@ The **`data-*`** [global attributes](/en-US/docs/Web/HTML/Global_attributes) for
 {{EmbedInteractiveExample("pages/tabbed/attribute-data.html","tabbed-standard")}}
 
 All such custom data are available via the {{domxref("HTMLElement")}} interface of the element the attribute is set on. The {{domxref("HTMLElement.dataset")}} property gives access to them.
-The `*` may be replaced by any name following [the production rule of XML names](https://www.w3.org/TR/REC-xml/#NT-Name) with the following restrictions:
+The `*` may be replaced by any name following [the production rule of XML names](https://www.w3.org/TR/REC-xml/#NT-Name) which includes the following restrictions:
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/data-_star_/index.md:15]:
```suggestion
The `*` may be replaced by any name following [the production rule of XML names](https://www.w3.org/TR/REC-xml/#NT-Name) which includes the following recommendations:
```

**Comment by sidewayss** [comment]
If you really want to get pedantic, I can change:
- All such custom data are available via the [HTMLElement](https://pr28686.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/HTMLElement) interface of the element the attribute is set on.

to
- All such custom data are available via the [HTMLElement](https://pr28686.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/HTMLElement) interface of the element on which the attribute is set.

File: files/en-us/web/html/global_attributes/data-_star_/index.md [hunk 3]
```diff
@@ -12,13 +12,13 @@ The **`data-*`** [global attributes](/en-US/docs/Web/HTML/Global_attributes) for
...
 
-Note that the {{domxref("HTMLElement.dataset")}} property is a {{domxref("DOMStringMap")}}, and the name of the custom data attribute _data-test-value_ will be accessible via `HTMLElement.dataset.testValue` (or by `HTMLElement.dataset["testValue"]`) as any dash (`U+002D`) is replaced by the capitalization of the next letter, converting the name to camelcase.
+In JavaScript, the {{domxref("HTMLElement.dataset")}} property is a {{domxref("DOMStringMap")}} containing all the custom data attributes for an {{domxref("HTMLElement")}}. For example, `element.getAttribute("data-test")` is equivalent to `element.dataset.test`. The `dataset` version of the attribute name removes the "data-" prefix and converts the rest of the name from snake-case to camelCase. For example, `data-test-abc` will be accessible as `HTMLElement.dataset.testAbc` (or by `HTMLElement.dataset["testAbc"]`). Non-alphabetic characters following a hyphen are not valid, for example `data-test-1` or `data--test`. If you plan to use {{domxref("HTMLElement.dataset")}}, the name can only include characters allowed in JavaScript property names (and hyphens, which will be removed).
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/data-_star_/index.md]:
```suggestion
These are just recommendations. If these naming recommendations are not followed, no errors will occur. The attributes will still be matched using CSS [attribute selectors](/en-US/docs/Web/CSS/Attribute_selectors), with the attribute being case insensitive and any attribute value being case-sensitive. Attributes not conforming to these three recommendations will also still be recognized by the  JavaScript {{domxref("HTMLElement.dataset")}} property and user-agents will include the attribute in the {{domxref("DOMStringMap")}} containing all the custom data attributes for an {{domxref("HTMLElement")}}.

If you plan to use {{domxref("HTMLElement.dataset")}}, the portion of the attribute name following the `data-` can only include characters allowed in JavaScript property names (and hyphens, which will be removed). The `dataset` version of the attribute name removes the "data-" prefix and converts the rest of the name from snake-case to camelCase. For example, `element.getAttribute("data-test")` is equivalent to `element.dataset.test` and `data-test-abc` will be accessible as `HTMLElement.dataset.testAbc` (or by `HTMLElement.dataset["testAbc"]`). Avoid non-alphabetic characters following a hyphen, such as `data-test-1` or `data--test`,  as they will not be recognized by {{domxref("HTMLElement.dataset")}}.
```

**Review by estelle — COMMENTED** [review]
i think we may have crossed paths again.

**Comment by estelle** [comment]
> i think we may have crossed paths again.

yup. I started reviewing before your last few comments.

**Comment by sidewayss** [comment]
No problem.  I've committed your suggestions except for the one that is marked as outdated.  Let me know if there are any further changes you'd like me to make.

File: files/en-us/web/html/global_attributes/data-_star_/index.md [hunk 4]
```diff
@@ -12,13 +12,13 @@ The **`data-*`** [global attributes](/en-US/docs/Web/HTML/Global_attributes) for
...
 
-Note that the {{domxref("HTMLElement.dataset")}} property is a {{domxref("DOMStringMap")}}, and the name of the custom data attribute _data-test-value_ will be accessible via `HTMLElement.dataset.testValue` (or by `HTMLElement.dataset["testValue"]`) as any dash (`U+002D`) is replaced by the capitalization of the next letter, converting the name to camelcase.
+In JavaScript, the {{domxref("HTMLElement.dataset")}} property is a {{domxref("DOMStringMap")}} containing all the custom data attributes for an {{domxref("HTMLElement")}}. For example, `element.getAttribute("data-test")` is equivalent to `element.dataset.test`. The `dataset` version of the attribute name removes the "data-" prefix and converts the rest of the name from snake-case to camelCase. For example, `data-test-abc` will be accessible as `element.dataset.testAbc` (or by `element.dataset["testAbc"]`). Non-alphabetic characters following a hyphen are not valid, for example `data-test-1` or `data--test`. If you plan to use {{domxref("HTMLElement.dataset")}}, the name can only include characters allowed in JavaScript property names (and hyphens, which will be removed).
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/data-_star_/index.md:21]:
```suggestion
These are recommendations. If these naming recommendations are not followed, no errors will occur. The attributes will still be matched using CSS [attribute selectors](/en-US/docs/Web/CSS/Attribute_selectors), with the attribute being case insensitive and any attribute value being case-sensitive. Attributes not conforming to these three recommendations will also still be recognized by the  JavaScript {{domxref("HTMLElement.dataset")}} property and user-agents will include the attribute in the {{domxref("DOMStringMap")}} containing all the custom data attributes for an {{domxref("HTMLElement")}}.

If you plan to use {{domxref("HTMLElement.dataset")}}, the portion of the attribute name following the `data-` can only include characters allowed in JavaScript property names (and hyphens, which will be removed). The `dataset` version of the attribute name removes the "data-" prefix and converts the rest of the name from snake-case to camelCase. For example, `element.getAttribute("data-test")` is equivalent to `element.dataset.test` and `data-test-abc` will be accessible as `HTMLElement.dataset.testAbc` (or by `HTMLElement.dataset["testAbc"]`). Avoid non-alphabetic characters following a hyphen, such as `data-test-1` or `data--test`,  as they will not be recognized by {{domxref("HTMLElement.dataset")}}.
```

**Review by estelle — APPROVED** [review]
thanks. Glad you found this issue and we were able to make this more accurate.

**Comment by sidewayss** [comment]
You're welcome. I'll be copying and pasting this into the SVG version of this page unless you say otherwise.  It's nice that the mdn-linter fixes things in the merge instead of just complaining like Prettier.  I had not seen that before.

**Pull request merged**